File: C/OpenMP_C_BASIC/example_5_OpenMP_Hello_World_Subroutine.c
/*  
 Compile
 $ gcc -g -Wall -fopenmp -o out example_5_OpenMP_Hello_World_Subroutine.c
Execute
 $ ./out <number of threads>
 */
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>   

void Hello(void);  /* Thread function */

/*main--------------------------------------------------------------------*/
int main(int argc, char* argv[]) {
    
   Hello();

   return 0; 
} 

/*-------------------------------------------------------------------
 * Function:    Hello
 * Purpose:     Thread function that prints message
 */
void Hello(void) {
    
  int nthreads, tid;
  
// fork a team of threads giving them their own copies of variables
  #pragma omp parallel private(nthreads, tid)
  {
// obtain thread number
     tid = omp_get_thread_num();
     printf("hello world from thread = %d\n", tid);

// only master thread does this
     if (tid == 0) {
        nthreads = omp_get_num_threads();
        printf("number of threads = %d\n", nthreads);
     }
  }// all threads join master thread and disband
} 

File: Fortran/OpenMP_Fortran_BASIC/example_1_Hello_World/example_5_OpenMP_Hello_World_Subroutine.f90
! $ gfortran -g -Wall -fopenmp -o out example_5_OpenMP_Hello_World_Subroutine.f90
! $ export OMP_NUM_THREADS=16
! $ ./out
!  hello world from thread =            2
!  hello world from thread =            5
!  hello world from thread =           11
!  hello world from thread =           10
!  hello world from thread =            6
!  hello world from thread =            8
!  hello world from thread =           14
!  hello world from thread =           12
!  hello world from thread =           15
!  hello world from thread =            4
!  hello world from thread =            3
!  hello world from thread =            7
!  hello world from thread =            1
!  hello world from thread =            9
!  hello world from thread =            0
!  number of threads =           16
!  hello world from thread =           13
! -------------------------------------------------------------------

program hello

  call hellofunc

end
! -------------------------------------------------------------------
subroutine hellofunc

  use omp_lib 
  integer :: nthreads, tid
  !integer :: nthreads, tid, omp_get_num_threads, omp_get_thread_num

! fork a team of threads giving them their own copies of variables
!$omp parallel private(nthreads, tid)

  ! obtain thread number
  tid = omp_get_thread_num()
  write(*,*) 'hello world from thread = ', tid

  ! only master thread does this
  if (tid .eq. 0) then
    nthreads = omp_get_num_threads()
    write(*,*) 'number of threads = ', nthreads
  end if
!$omp end parallel
! all threads join master thread and disband

end 
